### PR TITLE
Not All Tests are the Same By Sapan Parikh

### PR DESCRIPTION
**Pull request by haroon-sheikh**

### CODE REVIEW

File: sapan-parikh/SAPAN_PARIKH.md [hunk 1]
```diff
@@ -0,0 +1,16 @@
+# Not All Tests are the Same
+
+I have gotten an opportunity to work on projects where I learned how automated testing is and that not all automated tests are the same. They differ from each other in at least two ways,speed of execution and risk of failure.
+
+While the testing pyramid is QA 101, it'salso one of the most ignored learning. One project that I worked on depended heavily on unit testing, while the other relied heavily on automated() testing. approaches have their shortcomings.
```
Comment (emyjamalian) [inline sapan-parikh/SAPAN_PARIKH.md:5]: QA 101, it's also (a space is missing)

Comment (emyjamalian) [inline sapan-parikh/SAPAN_PARIKH.md:5]: unit testing is also automated testing though, do you maybe mean end to end testing?

File: sapan-parikh/SAPAN_PARIKH.md [hunk 2]
```diff
@@ -0,0 +1,16 @@
+# Not All Tests are the Same
+
+I have gotten an opportunity to work on projects where I learned how automated testing is and that not all automated tests are the same. They differ from each other in at least two ways,speed of execution and risk of failure.
```
Comment (melissajanefisher) [inline sapan-parikh/SAPAN_PARIKH.md:3]: Suggestion: I have worked on projects where I learned how automated tests are not all the same..

File: sapan-parikh/SAPAN_PARIKH.md (hunk 1 above)
Comment (flynnbops) [reply sapan-parikh/SAPAN_PARIKH.md:5]: Worth adding a link.  Things that we think are basic / 101 aren't always the case. How about something like: `While the [testing pyramid])(LINK) is a great model`

Comment (flynnbops) [reply sapan-parikh/SAPAN_PARIKH.md:5]: Similar thought as emyjamalian. End to end testing? UI testing? etc

File: sapan-parikh/SAPAN_PARIKH.md [hunk 3]
```diff
@@ -0,0 +1,16 @@
...
+
+While the testing pyramid is QA 101, it'salso one of the most ignored learning. One project that I worked on depended heavily on unit testing, while the other relied heavily on automated() testing. approaches have their shortcomings.
+
+Approach , which depended solely on unit testing of backend code, had many false positives. In this scenario, the test cases passed all the time. , the software in production did not work. This false assurance could be because of a lack of Javascript tests or 3rd party integration errors because of the limited unit tests.
```
Comment (flynnbops) [inline sapan-parikh/SAPAN_PARIKH.md:7]: How about `The first approach...` leads into your account.

File: sapan-parikh/SAPAN_PARIKH.md [hunk 4]
```diff
@@ -0,0 +1,16 @@
...
+
+On the other side, the second approach had precisely the opposite issue. This approach had lots of false negatives. Here many test cases failed, but things were not broken to scale indicated. For instance, 100 broken test cases, or a few percent of test cases always break due to timeouts or other issues not related to the code in production.
```
Comment (flynnbops) [inline sapan-parikh/SAPAN_PARIKH.md:9]: How about "The second  approach ..."  i.e. remove `On the other side`

File: sapan-parikh/SAPAN_PARIKH.md [hunk 5]
```diff
@@ -0,0 +1,16 @@
...
+
+This discussion brings us back to the testing pyramid. Any organization or project'stesting strategy should include unit tests, integration tests, Selenium tests, manual tests, etc. Such atest strategy can tell us exactly the problems arewith minimum effort and time.This is how tests differ in size and speed, but they also differ in their impact and risk.It makes sense that when your project is small, you run all the tests, but as it grows, executing all tests may take hours or days. A simple strategy at such times is to distribute your test cases in a quadrant of the likelihoodof failure and impact of failure. A test case with a higher possibility of failure and a high impactisthe kindthat should be prioritized.
```
Comment (flynnbops) [inline sapan-parikh/SAPAN_PARIKH.md:11]: Missing spaces between a few words.

Comment (flynnbops) [inline sapan-parikh/SAPAN_PARIKH.md:11]: Like the point you're making about tests at different levels. Do you need to specify "Selenium"? Your point appears to be "A test strategy should include tests at different test levels, e.g. unit, integration, UI, etc"

Comment (flynnbops) [reply sapan-parikh/SAPAN_PARIKH.md:11]: Is `quadrant of the likelihood of failure and impact of failure` a model you've seen. Sounds like a link could be useful

File: sapan-parikh/SAPAN_PARIKH.md [hunk 6]
```diff
@@ -0,0 +1,16 @@
...
+
+Slow running test cases can be given score between 1 to 5 for their possibility of failure and impact of failure. Based on historical data, one can find the probability of the test case failing again. mpact can be based on the possible damage if feature work. The product of both these scores should give us a risk score. Which typically looks like the following:
```
Comment (flynnbops) [inline sapan-parikh/SAPAN_PARIKH.md:13]: Is this score link to the quadrants mentioned above?

Comment (flynnbops) [reply sapan-parikh/SAPAN_PARIKH.md:13]: Typo `mpact` should be `impact`

File: sapan-parikh/SAPAN_PARIKH.md [hunk 7]
```diff
@@ -0,0 +1,16 @@
...
+
+![Slow running](sapan-parikh/slowrunning.png "Slow Running")
```
Comment (flynnbops) [inline sapan-parikh/SAPAN_PARIKH.md:15]: The MR doesn't include a valid image as far as I can see. Could be on my end only.

**Review by flynnbops — COMMENTED** [review]
Thanks for sharing. Sounds like you've benefitted from using models to help your testing approach. Some links could be useful for the readers

**Comment by sapanparikh18** [comment]
Hi Team, thanks a lot for creating the PR and putting so many comments here. Because of some kind of formatting or copy-paste issue, there are so many issues that could have been avoided and I apologize for it. I will be submitting a new draft soon

**Comment by haroon-sheikh** [comment]
Moved to https://github.com/testingstories/testingstories/pull/14.

**Pull request closed without merging**